### PR TITLE
Sections

### PR DESCRIPTION
**Pull request by jlconlin**
This is the beginning of what a Section is, as well as the definition of Section<3>

### CODE REVIEW

File: implementation/SectionBase.hpp.in [hunk 1]
```diff
@@ -3,5 +3,18 @@
...
 
+class ENDFtk::implementation::SectionBase{
+public:
+  SectionBase( const HEAD& head );
+
+  int MT() const { return MT_; }
+  int ZA() const { return ZA_; }
+  double atomicWeightRatio() const { return atomicWeightRatio_; }
+
+  const int MT_;
```
Comment (apmccartney) [inline implementation/SectionBase.hpp.in]: Why enforce that a section's member variables are always `const`?

File: implementation/SectionBase.hpp.in [hunk 2]
```diff
@@ -3,5 +3,18 @@
 
 #include "ENDFtk.hpp"
 
+class ENDFtk::implementation::SectionBase{
+public:
+  SectionBase( const HEAD& head );
+
+  int MT() const { return MT_; }
```
Comment (apmccartney) [inline implementation/SectionBase.hpp.in]: Mutable getters please, e.g.
```cpp
int& MT(){ return this->MT_; }
```

File: implementation/SectionBase.hpp.in [hunk 3]
```diff
@@ -3,5 +3,18 @@
...
+  int ZA() const { return ZA_; }
+  double atomicWeightRatio() const { return atomicWeightRatio_; }
```
Comment (apmccartney) [inline implementation/SectionBase.hpp.in]: Good name choice.
I'd also like to also support the manual acronymn `AWR`

File: implementation/SkipRecord/test/SkipRecord.test.cpp [hunk 4]
```diff
@@ -13,7 +13,7 @@ int main( int argc, const char* argv[] ){
   return result;
 }
 
-SCENARIO( "TextRecord Tests", "[ENDFtk], [TextRecord]" ){
+SCENARIO( "SkipRecord Tests", "[ENDFtk], [SkipRecord]" ){
```
Comment (apmccartney) [inline implementation/SkipRecord/test/SkipRecord.test.cpp]: good catch

File: implementation/HeadRecord.hpp.in [hunk 5]
```diff
@@ -14,6 +14,10 @@ public:
               int MAT, int MF, int MT ) :
   StructureDivision( C1, C2, L1, L2, N1, N2, MAT, MF, MT ) {}
   
+  template< typename Iterator >
+  HeadRecord( Iterator& it, const Iterator& end, long& lineNumber )
+    : StructureDivision(it, end, lineNumber) { /* Do nothing */ }
+
```
Comment (apmccartney) [inline implementation/HeadRecord.hpp.in]: promise to only use it in testing?

File: implementation/SectionBase.hpp.in [hunk 6]
```diff
@@ -16,6 +16,23 @@ public:
   int ZA() const { return ZA_; }
   double atomicWeightRatio() const { return atomicWeightRatio_; }
 
+  template<typename Iterator>
+  void readSEND(int MF, Iterator& begin, const Iterator& end, long LN){
```
Comment (apmccartney) [inline implementation/SectionBase.hpp.in]: The file number, i.e. MF, need not be passed. It's a template parameter.

The MAT number should be passed for error logging. The MT and MF are not sufficient to uniquely identify a given section on a tape. Providing the line number in the error message would be helpful as well.

Can this function be called in the in the initialization list of the ctor?

File: implementation/SectionBase.hpp.in [hunk 7]
```diff
@@ -3,5 +3,18 @@
 
 #include "ENDFtk.hpp"
 
+class ENDFtk::implementation::SectionBase{
+public:
+  SectionBase( const HEAD& head );
```
Comment (apmccartney) [inline implementation/SectionBase.hpp.in]: Consider accepting MAT and MF numbers for verification

File: implementation/SectionBase/Section3.hpp.in [hunk 8]
```diff
@@ -4,8 +4,11 @@
 #include "ENDFtk/implementation/SectionBase.hpp"
 #include "ENDFtk/implementation/UnivariateTabulation.hpp"
 
+namespace ENDFtk{
```
Comment (apmccartney) [inline implementation/SectionBase/Section3.hpp.in]: Isn't it weird clang doesn't complain about this?

File: implementation/SectionBase/Section3.hpp.in [hunk 9]
```diff
@@ -2,20 +2,39 @@
...
+    LOG(INFO) << "Trouble constructing MF=3 Section.";
+    throw e;
+  }
     
   using SectionBase::MT;
   using SectionBase::ZA;
   using SectionBase::atomicWeightRatio;
 
+  ENDFtk::implementation::UnivariateTabulation TAB1_;
+
```
Comment (apmccartney) [inline implementation/SectionBase/Section3.hpp.in]: Mutable and immutable getters for `TAB1_` please

File: implementation/SectionBase/Section3.hpp.in [hunk 10]
```diff
@@ -2,20 +2,39 @@
...
+    try: SectionBase(head),
+         TAB1_(begin, end, lineNumber, head.MAT(), head.MF(), head.MT() )
+  { 
+    try{
+      auto SEND = ENDFtk::implementation::SEND(head.MAT(), MT_ );
+      if( not SEND.isSend() ){
+        LOG(ERROR) << "No SEND Record at end of MF=3, MT=" << MT_ << ".";
+        throw std::exception();
+      }
+    } catch ( std::exception& e){
```
Comment (apmccartney) [inline implementation/SectionBase/Section3.hpp.in]: There are additional constraints on the `Tab1` record here than the general case.
Can you verify the yvalues of the tab1 are non-negative?

File: implementation/SectionBase.hpp.in (hunk 6 above)
Comment (jlconlin) [inline implementation/SectionBase.hpp.in]: make LN a reference

File: implementation/SectionBase.hpp.in [hunk 11]
```diff
@@ -10,32 +10,41 @@
  */
 class ENDFtk::implementation::SectionBase{
 public:
-  SectionBase( const HEAD& head );
+  SectionBase( const HEAD& head, int MAT, int MF );
 
   int MT() const { return MT_; }
+  const int& MT() { return MT_; }
```
Comment (apmccartney) [inline implementation/SectionBase.hpp.in]: not mutable

File: implementation/SectionBase.hpp.in [hunk 12]
```diff
@@ -10,32 +10,41 @@
...
-  void readSEND(int MF, Iterator& begin, const Iterator& end, long LN){
+  void readSEND(Iterator& begin, const Iterator& end, long& LN, 
+                int MAT, int MF ){
     try{
       auto SEND = ENDFtk::implementation::StructureDivision(
           begin, end, LN);
       if( not SEND.isSend() ){
-        LOG(ERROR) << "No SEND Record at end of MF=" 
-                   << MF << ", MT=" << MT_ << ".";
+        LOG(ERROR) << "No SEND Record at end of MAT=" << MAT
```
Comment (apmccartney) [inline implementation/SectionBase.hpp.in]: Message has a lot of redundant information with follow up info log

File: implementation/SectionBase.hpp.in [hunk 13]
```diff
@@ -10,32 +10,41 @@
...
+                   << ", MF=" << MF << ", MT=" << MT_ << ".";
         throw std::exception();
       }
```
Comment (apmccartney) [inline implementation/SectionBase.hpp.in]: No MAT or MF verification

File: implementation/SectionBase/Section3.hpp.in [hunk 14]
```diff
@@ -7,16 +7,42 @@
...
-         TAB1_(begin, end, lineNumber, head.MAT(), head.MF(), head.MT() )
+  Section(HEAD& head, Iterator& begin, const Iterator& end, long& lineNumber,
+          int MAT)
+    try: SectionBase(head, MAT, 3),
+         TAB1_(begin, end, lineNumber, head.MAT(), 3, head.MT() )
   { 
-    readSEND(3, begin, end, lineNumber);
+
+    auto ltzero = [](auto& y) { return y < 0.0; };
+    std::find_if(TAB1_.yBegin(), TAB1_.yEnd(), ltzero);
```
Comment (apmccartney) [inline implementation/SectionBase/Section3.hpp.in]: This line does nothing

**Comment by apmccartney** [comment]
Approved!

**Pull request merged**